google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,474 @@
+      return make_ready_future(finish_response);
+    });
+  }
+
+ private:
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  void OnFailure(Status status) {
+    AttemptRetry(status, retry_policy_prototype_->clone(),
+                 backoff_policy_prototype_->clone());
+  }
+
+  void FinishOnStreamFail(
+      std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+          stream) {
+    stream->Finish().then([this](future<Status> finish_status) {
+      OnFailure(finish_status.get());
+    });
+  }
+
+  void SetReadWriteFutures() {
+    if (read_reinit_done_.has_value()) {
+      // want to set member variable to nullopt before setting future returned
+      // to caller
+      auto read_promise = read_reinit_done_;
+      read_reinit_done_ = absl::optional();
+      read_promise->set_value();
+    }
+    if (write_reinit_done_.has_value()) {
+      auto write_promise = write_reinit_done_;
+      write_reinit_done_ = absl::optional();
+      write_promise->set_value();
+    }
+  }
+
+  void AttemptRetry(Status status, std::shared_ptr<RetryPolicy> retry_policy,
+                    std::shared_ptr<BackoffPolicy> backoff_policy) {
+    bool shutdown;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      shutdown = stream_state_ == State::kShutdown;
+    }
+    if (!shutdown && !retry_policy->IsExhausted() &&
+        retry_policy->OnFailure(status)) {
+      sleeper_(backoff_policy->OnCompletion())
+          .then([this, retry_policy, backoff_policy](future<void> sleep_fu) {
+            Initialize();
+          });
+      return;
+    }
+
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      stream_state_ = State::kShutdown;
+    }
+    status_promise_.set_value(kPermanentErrorStatus);
+    SetReadWriteFutures();
+  }
+
+  void Initialize(std::shared_ptr<RetryPolicy> retry_policy,
+                  std::shared_ptr<BackoffPolicy> backoff_policy) {
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+        potential_stream;
+    future<bool> start_future;
+
+    potential_stream = stream_factory_();
+    start_future = potential_stream->Start();
+
+    auto start_future_result = start_future.then(
+        [this, potential_stream =
+                   std::move(potential_stream)](future<bool> start_future) {
+          if (!start_future.get()) {
+            {
+              std::lock_guard<std::mutex> g{mu_};
+              stream_state_ = State::kRetrying;
+            }
+
+            FinishOnStreamFail(std::move(potential_stream));
+
+            // return this so the below continuation knows that it failed at the
+            // `Start` step
+            return make_ready_future(
+                StatusOr<std::unique_ptr<
+                    AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>(
+                    kStartErrorStatus));
+          }
+          return initializer_(std::move(potential_stream));
+        });
+
+    start_future_result.then(
```

Comment:
I think if you bind this to a member variable (retry_future_?) you can block the shutdown future returning until there are no retries in progress. You'd have to make the AttemptRetry method return a void future though

Reply:
Done